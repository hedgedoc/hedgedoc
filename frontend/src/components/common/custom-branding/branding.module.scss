/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

.regular-size {
  height: 50px;
}

.inline-size {
  height: 30px;
}
