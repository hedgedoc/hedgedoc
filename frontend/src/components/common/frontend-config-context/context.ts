/*
 * SPDX-FileCopyrightText: 2025 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { FrontendConfigDto } from '@hedgedoc/commons'
import { createContext } from 'react'

export const frontendConfigContext = createContext<FrontendConfigDto | undefined>(undefined)
