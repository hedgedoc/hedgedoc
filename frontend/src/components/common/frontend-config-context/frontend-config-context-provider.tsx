'use client'
/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { FrontendConfig } from '../../../api/config/types'
import { frontendConfigContext } from './context'
import type { PropsWithChildren } from 'react'
import React from 'react'

interface FrontendConfigContextProviderProps extends PropsWithChildren {
  config?: FrontendConfig
}

/**
 * Provides the given frontend configuration in a context or renders an error message otherwise.
 *
 * @param config the frontend config to provoide
 * @param children the react elements to show if the config is valid
 */
export const FrontendConfigContextProvider: React.FC<FrontendConfigContextProviderProps> = ({ config, children }) => {
  return config === undefined ? (
    <span className={'text-white bg-dark'}>No frontend config received! Please check the server log.</span>
  ) : (
    <frontendConfigContext.Provider value={config}>{children}</frontendConfigContext.Provider>
  )
}
