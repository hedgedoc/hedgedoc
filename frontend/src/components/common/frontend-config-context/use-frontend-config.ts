/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { FrontendConfig } from '../../../api/config/types'
import { frontendConfigContext } from './context'
import { Optional } from '@mrdrogdrog/optional'
import { useContext } from 'react'

/**
 * Retrieves the current frontend config from the next react context.
 */
export const useFrontendConfig = (): FrontendConfig => {
  return Optional.ofNullable(useContext(frontendConfigContext)).orElseThrow(
    () => new Error('No frontend config context found. Did you forget to use the provider component?')
  )
}
