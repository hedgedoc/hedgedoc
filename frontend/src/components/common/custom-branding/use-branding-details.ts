/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { BrandingConfig } from '../../../api/config/types'
import { useFrontendConfig } from '../frontend-config-context/use-frontend-config'
import { useMemo } from 'react'

/**
 * Extracts the branding from the config.
 *
 * @return the branding configuration or null if no branding has been configured
 */
export const useBrandingDetails = (): null | BrandingConfig => {
  const branding = useFrontendConfig().branding

  return useMemo(() => {
    return !branding.name && !branding.logo ? null : branding
  }, [branding])
}
