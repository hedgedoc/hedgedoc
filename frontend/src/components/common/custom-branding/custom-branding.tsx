/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { concatCssClasses } from '../../../utils/concat-css-classes'
import styles from './branding.module.scss'
import { useBrandingDetails } from './use-branding-details'
import React, { useMemo } from 'react'

export interface BrandingProps {
  inline?: boolean
}

/**
 * Show the branding of the HedgeDoc instance.
 * This branding can either be a text, a logo or both (in that case the text is used as the title and alt of the image).
 *
 * @param inline If the logo should be using the inline-size or the regular-size css class.
 */
export const CustomBranding: React.FC<BrandingProps> = ({ inline = false }) => {
  const branding = useBrandingDetails()

  const className = useMemo(
    () =>
      concatCssClasses({
        [styles['regular-size']]: !inline,
        [styles['inline-logo']]: inline && branding && !!branding.logo,
        [styles['inline-text']]: inline && branding && !branding.logo
      }),
    [inline, branding]
  )

  if (!branding) {
    return null
  } else if (branding.logo) {
    return (
      /* eslint-disable-next-line @next/next/no-img-element */
      <img src={branding.logo} alt={branding.name} title={branding.name} className={className} />
    )
  } else {
    return <span className={className}>{branding.name}</span>
  }
}
