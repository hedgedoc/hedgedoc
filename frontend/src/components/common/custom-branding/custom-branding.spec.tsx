/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { FrontendConfig } from '../../../api/config/types'
import * as UseFrontendConfigMock from '../frontend-config-context/use-frontend-config'
import { CustomBranding } from './custom-branding'
import { render } from '@testing-library/react'
import { Mock } from 'ts-mockery'

jest.mock('../frontend-config-context/use-frontend-config')

describe('custom branding', () => {
  const mockFrontendConfigHook = (logo?: string, name?: string) => {
    jest
      .spyOn(UseFrontendConfigMock, 'useFrontendConfig')
      .mockReturnValue(Mock.of<FrontendConfig>({ branding: { logo, name } }))
  }

  it("doesn't show anything if no branding is defined", () => {
    mockFrontendConfigHook()
    const view = render(<CustomBranding />)
    expect(view.container).toMatchSnapshot()
  })

  describe.each([false, true, undefined])('with inline=%s', (inline) => {
    it('shows an image if branding logo is defined', () => {
      mockFrontendConfigHook('mockBrandingUrl')
      const view = render(<CustomBranding inline={inline} />)
      expect(view.container).toMatchSnapshot()
    })

    it('shows an text if branding text is defined', () => {
      mockFrontendConfigHook(undefined, 'mockedBranding')
      const view = render(<CustomBranding inline={inline} />)
      expect(view.container).toMatchSnapshot()
    })

    it('will prefer the logo over the text', () => {
      mockFrontendConfigHook('mockBrandingUrl', 'mockedBranding')
      const view = render(<CustomBranding inline={inline} />)
      expect(view.container).toMatchSnapshot()
    })
  })
})
