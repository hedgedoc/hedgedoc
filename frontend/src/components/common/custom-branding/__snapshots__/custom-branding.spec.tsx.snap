// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom branding doesn't show anything if no branding is defined 1`] = `<div />`;

exports[`custom branding with inline=false shows an image if branding logo is defined 1`] = `
<div>
  <img
    class="regular-size"
    src="mockBrandingUrl"
  />
</div>
`;

exports[`custom branding with inline=false shows an text if branding text is defined 1`] = `
<div>
  <span
    class="regular-size"
  >
    mockedBranding
  </span>
</div>
`;

exports[`custom branding with inline=false will prefer the logo over the text 1`] = `
<div>
  <img
    alt="mockedBranding"
    class="regular-size"
    src="mockBrandingUrl"
    title="mockedBranding"
  />
</div>
`;

exports[`custom branding with inline=true shows an image if branding logo is defined 1`] = `
<div>
  <img
    class="inline-logo"
    src="mockBrandingUrl"
  />
</div>
`;

exports[`custom branding with inline=true shows an text if branding text is defined 1`] = `
<div>
  <span
    class="inline-text"
  >
    mockedBranding
  </span>
</div>
`;

exports[`custom branding with inline=true will prefer the logo over the text 1`] = `
<div>
  <img
    alt="mockedBranding"
    class="inline-logo"
    src="mockBrandingUrl"
    title="mockedBranding"
  />
</div>
`;

exports[`custom branding with inline=undefined shows an image if branding logo is defined 1`] = `
<div>
  <img
    class="regular-size"
    src="mockBrandingUrl"
  />
</div>
`;

exports[`custom branding with inline=undefined shows an text if branding text is defined 1`] = `
<div>
  <span
    class="regular-size"
  >
    mockedBranding
  </span>
</div>
`;

exports[`custom branding with inline=undefined will prefer the logo over the text 1`] = `
<div>
  <img
    alt="mockedBranding"
    class="regular-size"
    src="mockBrandingUrl"
    title="mockedBranding"
  />
</div>
`;
